File: firebase/functions/lib/utils/send_email_client.dart
import 'package:firebase_admin_interop/firebase_admin_interop.dart';
import 'infra/firestore_utils.dart';
import 'package:data_models/cloud_functions/requests.dart';
import 'package:quiver/iterables.dart';

SendEmailClient sendEmailClient = SendEmailClient();

class SendEmailClient {
  Future<void> sendEmail(
    SendGridEmail email, {
    Transaction? transaction,
  }) async {
    final newDocument = firestore.collection('sendgridmail').document();

    final newData =
        DocumentData.fromMap(firestoreUtils.toFirestoreJson(email.toJson()));
    if (transaction != null) {
      transaction.create(newDocument, newData);
    } else {
      await newDocument.setData(newData);
    }
  }

  Future<void> sendEmails(List<SendGridEmail> emails) async {
    await Future.wait(
      partition(emails, 500).map((sublist) {
        final batch = firestore.batch();
        sublist.map((email) {
          final doc = firestore.collection('sendgridmail').document();
          batch.setData(
            doc,
            DocumentData.fromMap(
              firestoreUtils.toFirestoreJson(email.toJson()),
            ),
          );
        }).toList();
        return batch.commit();
      }).toList(),
    );
  }
}

File: firebase/functions/src/utils/send_email_client.ts
import { firestore, firestoreUtils } from '../utils/infra/firestore_utils';
import { SendGridEmail } from '../types';

export class SendEmailClient {
  async sendEmail(email: SendGridEmail, opts?: { transaction?: FirebaseFirestore.Transaction }): Promise<void> {
    const newDocument = firestore.collection('sendgridmail').doc();
    const newData = firestoreUtils.toFirestoreJson(serializeSendGridEmail(email));

    if (opts?.transaction) {
      opts.transaction.set(newDocument, newData);
    } else {
      await newDocument.set(newData);
    }
  }

  async sendEmails(emails: SendGridEmail[]): Promise<void> {
    // Split into batches of 500
    const chunks: SendGridEmail[][] = [];
    for (let i = 0; i < emails.length; i += 500) {
      chunks.push(emails.slice(i, i + 500));
    }

    await Promise.all(
      chunks.map((sublist) => {
        const batch = firestore.batch();
        for (const email of sublist) {
          const doc = firestore.collection('sendgridmail').doc();
          batch.set(doc, firestoreUtils.toFirestoreJson(serializeSendGridEmail(email)));
        }
        return batch.commit();
      })
    );
  }
}

function serializeSendGridEmail(email: SendGridEmail): Record<string, unknown> {
  return {
    to: email.to,
    from: email.from,
    message: {
      subject: email.message.subject,
      html: email.message.html,
      attachments: email.message.attachments,
    },
  };
}

export function setSendEmailClient(instance: SendEmailClient): void {
  sendEmailClient = instance;
}

export let sendEmailClient = new SendEmailClient();
